File: PIGBANK2.py
for _ in range(int(input())):
    pew, pfw = tuple(map(int, input().split()))
    coins = tuple(map(lambda _: tuple(map(int, input().split())), range(int(input()))))
    m, s = {0:0}, "This is impossible."
    for i in range(1, pfw - pew + 1):
        for p, w in coins:
            if i - w in m:
                m[i] = min(m.get(i, m[i - w] + p), m[i - w] + p)
    if pfw - pew in m:
        s = "The minimum amount of money in the piggy-bank is {}.".format(m[pfw - pew])
    print(s)

File: PIGBANK2.lua
for _ = 1, io.read("*n") do
    pew, pfw, cc = io.read("*n", "*n", "*n")
    coins, m, s = {}, {}, "This is impossible."
    for _ = 1, cc do
        coins[#coins + 1] = table.pack(io.read("*n", "*n"))
    end
    m[0] = 0
    for i = 1, pfw - pew do
        for _, c in ipairs(coins) do
            w, p = m[i - c[2]], c[1]
            m[i] = w and math.min(w + p, m[i] or w + p) or m[i]
        end
    end
    if m[pfw - pew] then
        s = string.format("The minimum amount of money in the piggy-bank is %d.", m[pfw - pew])
    end
    print(s)
end
